### ISSUE TITLE
optionalorrequired can't report definition on root

### ISSUE THREAD

**Issue by sivchari** [kubernetes-sigs/kube-api-linter#51]
Currently, optionalorrequired linter can't report if the type that field has is defined type and marker is set on the definition side, not Field side.
For example, the following case pass

```go
type Spec struct {
	Enum Enum `json:"enum"`
}

// +required
// +kubebuilder:validation:Eum=foo;bar;baz
type Enum string
```

We should improve shared-iterator about markers. Maybe, this improvement affects other linters that uses marker analyzer.

**Comment by JoelSpeed** [comment]
So is this saying that the linter will report the Enum field as missing the marker? I would argue, specifically for the case of optional vs required, that adding a required tag to a type marker is not supposed to be done/supported so I'd rather we actually add a check that reports and removes optional/required markers from type definitions

**Comment by sivchari** [comment]
> So is this saying that the linter will report the Enum field as missing the marker?

I was thinking that the linter should report the Spec.Enum as missing marker or report Enum type as that type have marker and suggest to remove from it and add it to the field.

> specifically for the case of optional vs required, that adding a required tag to a type marker is not supposed to be done/supported

Is it recommendation ? Because the same spec is generated whether the CRD has the marker in the type declaration or in the field.

> so I'd rather we actually add a check that reports and removes optional/required markers from type definitions

It sounds good to me. I'll open the PR corresponding to it later.

**Comment by JoelSpeed** [comment]
> I was thinking that the linter should report the Spec.Enum as missing marker or report Enum type as that type have marker and suggest to remove from it and add it to the field.

Yes I think I agree, can you raise a PR to add a test case and we can work from there?

> Is it recommendation ? Because the same spec is generated whether the CRD has the marker in the type declaration or in the field.

Ok, so yes, you _can_ do this, but I think in general when adding markers to types, they should be things that are always true of the type. What then if you wanted to add a field using this type that was then required? Then there would be a conflict

IMO, we should prevent people from doing this since being able to do so is likely to lead to confusion. I wouldn't even make this configurable, but in theory, we could make it configurable in the future if someone decided that it would be better for them

**Comment by sivchari** [comment]
/close
resolved
